### ISSUE TITLE
Suggest adding driverVersion property on events emitted by driver

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#8069]
Example:

 Office_Fluid_FluidRuntime_Error
| where Data_error contains "TooManyRetries"
| summarize count() by Data_driverVersion, Data_eventName

This emits fluid:telemetry:DeltaManager:GetDeltas_Exception but very hard to say what version of driver was used to generate this error, as event comes from loader.
We can do joins with other events from same session to get this data, but I wonder if driver should just stamp version on all errors it raises?

See createOdspLogger() how we do it for all driver events today.

**Comment by markfields** [comment]
Updated the name ("events" => "errors")

**Comment by markfields** [comment]
Moving to Jan and adding design-required label.

Possible approaches:
* Support global props that propagate through whole logger graph (see #8346)
* Update error base classes in ODSP driver to include pkgVersion property -- BUT that code is in odsp-doclib-utils package, which could diverge in versioning from odsp-driver package (where `createOdspLogger` is).
* Pass in `driverVersion` as constructor arg when creating errors (simple but yuck)
* Update `IDocumentService` interface to include a `version` somewhere and container can add that value as `driverVersion` on its logger
* Not sure about this one... but createDocumentService takes the logger, there could be a way for it to shove `driverVersion` on there itself (and not just on the ChildLogger(s) it creates)

@vladsud, @anthony-murphy  -- thoughts?

**Comment by vladsud** [comment]
I actually like the direction of error throwing by a layer (like driver) containing enough info to clearly identify circumstance of situation, i.e. including driver version that would get logged (with all other properties).
While I'd love to see a world where every event contains loader/driver/runtime versions, it's not feasible, especially if number of layers increase and some layers are loaded on demand. So we will need to find ways to put info in right places and do various tricks (like query joins) to stitch the picture together.

I do not see a difference (in versions) between driver package and odsp-doclib-utils (they are one and the same from versioning perspective), so base classes including version, or something like createOdspNetworkError() filling in version is fine. I do not think latter would survive the test of time (as I believe half of code paths landing there should not be using it - see #8258). But helper method to annotate error before its thrown / returned or baking it into ctors of errors seems fine.

**Comment by markfields** [comment]
> I do not see a difference (in versions) between driver package and odsp-doclib-utils (they are one and the same from versioning perspective)

@heliocliu, @anthony-murphy -- When you peer into the future of how our packages are versioned/released, does it seem reasonable to assume that we can lock `@fluidframework/odsp-doclib-utils` and `@fluidframework/odsp-driver` packages to be versioned together?

**Comment by heliocliu** [comment]
Seems reasonable to me (I think they're locked together in client?).  The package group locking is really only limited by dependency issues I think.

**Comment by anthony-murphy** [comment]
no i don't believe so. some of our partners dynamically load the driver itself, but i think they may also statically (npm) reference odsp-doclib-utils

**Comment by anthony-murphy** [comment]
we'll they can probably be versioned and released together, but anything they share will need to support cross versions. in terms of driverVersion, it can probably be the same number, as if they are different then the driver should have its own and matching version of odsp-doclib-utils

**Comment by markfields** [comment]
Ok cool.  Removing design-required label and leaving in Jan to execute on.

I'm realizing that this whole concept of "driverVersion" has a lot of assumptions built-in in terms of how the packages are consumed and bundled by our partners.  @heliocliu I wonder if we should formalize a concept of groups of packages where the code is assuming that package A will be bundled with package B of the same version?  At least in terms of guidance for people consuming our packages.

**Comment by vladsud** [comment]
I'd love to see some pragmatic solution here soon :) Yes, there are multiple packages, but I'm not sure we have to look for absolutely correct solution that takes into account future possibilities of shipping packages with different versions.
Clearly, we do not do that when reporting runtimeVersion, even though we do support mix & match of container runtime and data store runtime with N/N-1 & N/N-2 configs in both directions.

**Comment by heliocliu** [comment]
That's definitely something we'll need and is part of the broader question of how we publish announcements of new releases.  I think the idea of two packages being bundled with the same version of each other is already captured with the existing group structure (e.g. "client" packages) and just needs to be listed as such in any release notes.  Packages that expect certain but different versions of each other (e.g. cross version compat) get a lot more complicated, and I don't have a great solution here.  E.g. if driver, runtime, and loader are all different groups and all release at different speeds, this results in an impossible to understand compatibility matrix with just 3 members supporting N±2.  We could provide some web tool to determine compatibility for you and/or enforce it through something like peer dependencies, but that doesn't seem like a great experience.

**Comment by anthony-murphy** [comment]
i think this is getting over complicated. whatever code is running can know its version, via our genver stuff. layers should just stamp their version on error and logs. won't that be enough? then it doesn't matter how things are composed

**Comment by vladsud** [comment]
Yep, simple is better here. We will have plenty of opportunities to make it complex if we need to :)

**Comment by markfields** [comment]
> I'd love to see some pragmatic solution here soon 

Yes, the plan is to do the simple thing this month.

> layers should just stamp their version on error and logs. won't that be enough?

I'm just pointing out that "layer" makes sense in our head, but I'm about to write code that christens the genver result of a single package as the version for that whole "layer", but I can imagine that assumption being broken one day.  e.g. loaderVersion used to be enough for driver code too but then we split them up.  Again. not trying to distract from the short-term plan here, just looking ahead while we're talking about this stuff.

**Comment by markfields** [comment]
I started working on this, but got sidetracked because ConnectionManager (in loader package) is creating Driver Errors. This is not the end of the world, but it sent me off way in another direction of inquiry today.  Plan will be to have driverVersion be required in the constructors still, and pass `""` from outside driver packages.
